desktop/sources/scripts/clock.js

```diff
@@ -73,8 +89,17 @@ function Clock (client) {
     }
   }
 
-  this.stop = function (msg = false) {
+  this.stop = function (msg = false, linkStop = false) {
     console.log('Clock', 'Stop')
+    console.log(this.isLinkEnabled, this.isPaused, linkStop)
+    if (this.isLinkEnabled && !this.isPaused && !linkStop) {
+      this.isPaused = true
+      this.clearTimer()
+      client.link.stop()
+      client.io.midi.allNotesOff()
+      client.io.midi.silence()
+      return
+    }
```

Comment:
Several observations here:

* Inconsistent console logging between `play()` and `stop()`
* Link handling is not using `isPuppet` —  not sure if it's good or bad, but could also be considered inconsistent: Orca is "puppeteered" in both cases, it's just the control source that's changing.
* Link handler is injecting itself in play/stop control flow and duplicates handling of things that are handled down below.

It would be cleaner and more maintainable to combine `isPuppet` and `isLinkEnabled` into a single variable (this also nicely reflects the fact that the behaviors are mutually exclusive) and then homogenize the control flow. something like:

```js
const sourceClock = 1
const sourceLink = 2
var puppetSource = null
...

this.stop = function (msg = false, linkStop = false) {
  if (this.isPaused === true) { return }
  this.isPaused = true  // always want this
  if (this.puppetSource === sourceClock) {
    console.warn('Clock', 'External Midi control') // nothing else to do
  } else if {this.puppetSource === sourceLink) {
    this.clearTimer()
    client.link.stop()
  } else {
    if (msg === true || client.io.midi.isClock) { client.io.midi.sendClockStop() }
    this.clearTimer()
  }
  // always want this
  client.io.midi.allNotesOff()
  client.io.midi.silence()
}	  
```

Reply:
Or even:

```js
this.stop = ... {
  if (!this.isPaused) {
    this.isPaused = true
    ...
    // do all the other things
  }
}

Reply:
I think you might be right with this idea, I was trying to put a first iteration that worked and tried to not screw up your previous work 😅 

I'll look into this and push a new version ASAP